integrationTests/ts/test.js

```diff
@@ -4,16 +4,56 @@
 
 const path = require('path');
 const childProcess = require('child_process');
+const assert = require('assert');
+const fs = require('fs');
+const glob = require('glob');
+const semver = require('semver');
 
 const { dependencies } = require('./package.json');
 
 const tsVersions = Object.keys(dependencies)
   .filter((pkg) => pkg.startsWith('typescript-'))
   .sort((a, b) => b.localeCompare(a));
 
+// To allow omitting certain code from older versions of TypeScript, we have a
+// "magic" comment syntax. We precede a block of code with:
+//
+//     /*! SEMVER_RANGE !*/
+//
+// replacing SEMVER_RANGE with a semver range spec, such as '>=3.2'; we
+// terminate the block of code with:
+//
+//     /*!!*/
+//
+// We will only include the code between these comments if the TypeScript
+// version being tested satisfies the semver range that was specified. NOTE: We
+// currently do not allow nesting of these blocks.
+const templates = glob.sync('./*.ts.template').map((filename) => {
+  const content = fs.readFileSync(filename, 'utf8');
+  const targetFilename = filename.replace(/\.template$/, '');
+  assert.notEqual(filename, targetFilename);
+  const writeFileSync = (version) => {
+    // Captures our magic comment syntax: `/*!(CAPTURE1)!*/(CAPTURE 2)/*!!*/`
+    const regex = /\/\*!([^!]+)!\*\/([\s\S]*?)\/\*!!\*\//g;
+    const finalContent = content.replace(regex, (_, versionRange, payload) => {
+      if (semver.satisfies(version, versionRange)) {
+        return payload;
+      } else {
+        return '';
+      }
+    });
+    fs.writeFileSync(targetFilename, finalContent);
+  };
+  return { writeFileSync };
+});
+
 for (const version of tsVersions) {
   console.log(`Testing on ${version} ...`);
 
+  for (const template of templates) {
+    template.writeFileSync(version);
+  }
+
```

Comment:
@IvanGoncharov I seek your guidance on this approach to allowing a block of code to be omitted when testing older TypeScript versions. It works; but it does mean the tests write to the filesystem, which makes me uncomfortable. It's a fairly simple implementation though, basically powered by a single regexp and the `semver` module.

Reply:
@benjie Thanks for trying this solution.
How about just have `index-3.2.ts` and `index.ts` cycle starts with `index.ts` since we start from newest versions first but if `fs.fileExistSync(`index-${version.replace('typescript-', '')}.ts`)` exist we will switch to this file for this and latter versions.

Reply:
That was my first thought; but I realised that keeping these files in sync would be challenging, and it increases the amount of code in the tests without offering much in the way of value. It also doesn't allow for more complex ranges like `>=3.2 <5` which may be required in future. "Turning off" a block of code seemed like the simplest solution from the point of view of writing the tests, at a slight cost of the test infrastructure being more complex.

Reply:
I suppose the more complex ranges are enabled by having a `4.9` file and another `3.1` file; but this still seems more complex/cumbersome than just removing a block.

Reply:
@benjie couple of points:
1. don't like having `glob` and `semver` as dependencies it means we need to update them constantly to prevent GitHub marking this repo as affected by yet another security vulnerability.
2. Version after next will drop all TS up to 3.7 (to support recursive types) so this is the temporary workaround to keep maintaining `15.x.x` for some time and possibly backport critical fixes.
3. I don't expect this file to be changed frequently since we will migrate our codebase to TS we will need it just as a sanity check and for very rare occasions where we can't test stuff in our codebase like this one.

So I will likely remove this mechanism in `16.0.0` anyway so no need to make it too complex or future proof just make it to require as low maintenance as possible.

Reply:
@benjie Let's not block this PR and keep it KISS just add test as a comment with `FIXME` and I will uncomment it in `16.0.0`.
How does it sounds?

Reply:
1. You already have a lot of dependencies on both of these modules according to `package-lock.json`, but I totally get where you're coming from
2. Great :+1: 
3. Okay.

I've reverted the commit that adds this testing support.